File: identhub/src/main/java/de/solarisbank/identhub/contract/adapter/DocumentAdapter.java
package de.solarisbank.identhub.contract.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.jakewharton.rxrelay2.PublishRelay;

import java.util.ArrayList;
import java.util.List;

import de.solarisbank.identhub.R;
import de.solarisbank.sdk.data.dto.DocumentDto;
import io.reactivex.Observable;

public class DocumentAdapter extends RecyclerView.Adapter<DocumentViewHolder> {
    private List<DocumentDto> documents = new ArrayList<>();

    private PublishRelay<DocumentDto> actionClickRelay = PublishRelay.create();

    public Observable<DocumentDto> getActionOnClickObservable() {
        return actionClickRelay.hide();
    }

    @NonNull
    @Override
    public DocumentViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View holderView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_document, parent, false);
        return new DocumentViewHolder(holderView);
    }

    @Override
    public void onBindViewHolder(@NonNull DocumentViewHolder holder, int position) {
        DocumentDto document = documents.get(position);
        holder.bind(document);
        holder.bindAction(view -> {
            actionClickRelay.accept(document);
        });
    }

    @Override
    public int getItemCount() {
        return documents.size();
    }

    public void clear() {
        this.documents.clear();
    }

    public void add(List<DocumentDto> documents) {
        this.documents.addAll(documents);
    }

    public List<DocumentDto> getItems() {
        return documents;
    }
}

File: identhub/src/main/java/de/solarisbank/identhub/contract/adapter/DocumentAdapter.kt
package de.solarisbank.identhub.contract.adapter

import androidx.recyclerview.widget.RecyclerView
import de.solarisbank.sdk.data.dto.DocumentDto
import com.jakewharton.rxrelay2.PublishRelay
import android.view.ViewGroup
import android.view.LayoutInflater
import de.solarisbank.identhub.R
import io.reactivex.Observable
import java.util.ArrayList

class DocumentAdapter : RecyclerView.Adapter<DocumentViewHolder>() {
    private val documents: MutableList<DocumentDto> = ArrayList()
    private val actionClickRelay = PublishRelay.create<DocumentDto>()
    val actionOnClickObservable: Observable<DocumentDto>
        get() = actionClickRelay.hide()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DocumentViewHolder {
        val holderView = LayoutInflater.from(parent.context)
            .inflate(R.layout.identhub_item_document, parent, false)
        return DocumentViewHolder(holderView)
    }

    override fun onBindViewHolder(holder: DocumentViewHolder, position: Int) {
        val document = documents[position]
        holder.bind(document) { actionClickRelay.accept(document) }
    }

    override fun getItemCount(): Int {
        return documents.size
    }

    fun clear() {
        documents.clear()
    }

    fun add(documents: List<DocumentDto>?) {
        this.documents.addAll(documents!!)
    }
}